packages/backend-lalr/data/HappyTemplate.hs

```diff
@@ -194,35 +221,147 @@ happyDropStk n  (x `HappyStk` xs) = happyDropStk MINUS(n,(1#::Happy_Int)) xs
 happyGoto nt j tk st =
    DEBUG_TRACE(", goto state " ++ show (Happy_GHC_Exts.I# new_state) ++ "\n")
    happyDoAction j tk new_state
-   where off       = happyAdjustOffset (indexShortOffAddr happyGotoOffsets st)
-         off_i     = PLUS(off, nt)
-         new_state = indexShortOffAddr happyTable off_i
+  where new_state = happyIndexGotoTable nt st
 
 -----------------------------------------------------------------------------
--- Error recovery (ERROR_TOK is the error token)
-
--- parse error if we are in recovery and we fail again
-happyFail explist ERROR_TOK tk old_st _ stk@(x `HappyStk` _) =
-     let i = GET_ERROR_TOKEN(x) in
---      trace "failing" $
-        happyError_ explist i tk
-
-{-  We don't need state discarding for our restricted implementation of
-    "error".  In fact, it can cause some bogus parses, so I've disabled it
-    for now --SDM
-
--- discard a state
-happyFail  ERROR_TOK tk old_st (HappyCons action sts)
-                               (saved_tok `HappyStk` _ `HappyStk` stk) =
---      trace ("discarding state, depth " ++ show (length stk))  $
-        happyDoAction ERROR_TOK tk action sts (saved_tok`HappyStk`stk)
--}
-
--- Enter error recovery: generate an error token,
---                       save the old token and carry on.
-happyFail explist i tk action sts stk =
--- trace "entering error recovery" $
-        happyDoAction ERROR_TOK tk action sts (MK_ERROR_TOKEN(i) `HappyStk` stk)
+-- Error recovery
+--
+-- When there is no applicable action for the current lookahead token `tk`,
+-- happy enters error recovery mode. It works in 2 phases:
+--
+--  1. Fixup: Try to see if there is an action for the error token (`errorTok`,
+--     which is ERROR_TOK). If there is, do *not* emit an error and pretend
+--     instead that an `errorTok` was inserted.
+--     When there is no `errorTok` action, call the error handler
+--     (e.g., `happyError`) with the resumption continuation `happyResume`.
+--  2. Error resumption mode: If the error handler wants to resume parsing in
+--     order to report multiple parse errors, it will call the resumption
+--     continuation (of result type `P (Maybe a)`).
+--     In the absence of the %resumptive declaration, this resumption will
+--     always (do a bit of work, and) `return Nothing`.
+--     In the presence of the %resumptive declaration, the grammar author
+--     can use the special `catch` terminal to declare where parsing should
+--     resume after an error.
```

Comment:
This comment by me is out of date. Note that we no longer have a `%resumptive` pragma; we only have `%error { abort } { report }`, and `report` gets passed the resumption (of type `P a`).

The default behavior without `%error` (calling `happyError`) or with the unary variant of `%error` is to call a user-supplied function wrapped in `happyReport` that ignores the resumption. The reporting function will likely panic and abort the parse (at least `happyFail` does this). See also the code generation for `callReportError`.

It's all a bit confusing due to multiple layers of indirection; perhaps we can simplify.

It is probably best to describe the expected behavior of `%error { abort } { report }` in conjunction with the `catch` token in the user's guide first. Building on that documentation, we can describe how resumption is implemented here.